api_proto_test.go

```diff
@@ -0,0 +1,400 @@
+				fmt.Printf("got diff: %s", diff)
+				return false
+			}
+			return true
+		}
+		if err := quick.Check(f, nil); err != nil {
+			t.Fatal(err)
+		}
+	})
+}
+
+func (*IndexMetadata) Generate(r *rand.Rand, size int) reflect.Value {
+	indexTime := time.Now().Add(time.Duration(r.Int63n(1000)) * time.Hour)
+	var i IndexMetadata
+	i.IndexFormatVersion = gen(i.IndexFormatVersion, r)
+	i.IndexFeatureVersion = gen(i.IndexFeatureVersion, r)
+	i.IndexMinReaderVersion = gen(i.IndexMinReaderVersion, r)
+	i.IndexTime = indexTime
+	i.PlainASCII = gen(i.PlainASCII, r)
+	i.LanguageMap = gen(i.LanguageMap, r)
+	i.ZoektVersion = gen(i.ZoektVersion, r)
+	i.ID = gen(i.ID, r)
+	return reflect.ValueOf(&i)
+}
+
+func (*Repository) Generate(rng *rand.Rand, size int) reflect.Value {
+	latestCommitDate := time.Now().Add(time.Duration(rng.Int63n(1000)) * time.Hour)
+	var r Repository
+	v := &Repository{
+		ID:                   gen(r.ID, rng),
+		Name:                 gen(r.Name, rng),
+		URL:                  gen(r.URL, rng),
+		Source:               gen(r.Source, rng),
+		Branches:             gen(r.Branches, rng),
+		SubRepoMap:           map[string]*Repository{},
+		CommitURLTemplate:    gen(r.CommitURLTemplate, rng),
+		FileURLTemplate:      gen(r.FileURLTemplate, rng),
+		LineFragmentTemplate: gen(r.LineFragmentTemplate, rng),
+		priority:             gen(r.priority, rng),
+		RawConfig:            gen(r.RawConfig, rng),
+		Rank:                 gen(r.Rank, rng),
+		IndexOptions:         gen(r.IndexOptions, rng),
+		HasSymbols:           gen(r.HasSymbols, rng),
+		Tombstone:            gen(r.Tombstone, rng),
+		LatestCommitDate:     latestCommitDate,
+		FileTombstones:       gen(r.FileTombstones, rng),
+	}
+	return reflect.ValueOf(v)
+}
+
+func (RepoListField) Generate(rng *rand.Rand, size int) reflect.Value {
+	switch rng.Int() % 3 {
+	case 0:
+		return reflect.ValueOf(RepoListField(RepoListFieldRepos))
+	case 1:
+		return reflect.ValueOf(RepoListField(RepoListFieldMinimal))
+	default:
+		return reflect.ValueOf(RepoListField(RepoListFieldReposMap))
+	}
+}
+
+func gen[T any](sample T, r *rand.Rand) T {
+	var t T
+	v, _ := quick.Value(reflect.TypeOf(t), r)
+	return v.Interface().(T)
+}
+
+// This is a real search result that is intended to be a reasonable representative
+// for serialization benchmarks.
+var (
+	//go:embed testdata/search_result_1.pb
+	exampleSearchResultBytes []byte
+
+	// The proto struct representation of the search result
+	exampleSearchResultProto = func() *v1.SearchResponse {
+		sr := new(v1.SearchResponse)
+		err := proto.Unmarshal(exampleSearchResultBytes, sr)
+		if err != nil {
+			panic(err)
+		}
+		return sr
+	}()
+
+	// The non-proto struct representation of the search result
+	exampleSearchResultGo = SearchResultFromProto(exampleSearchResultProto)
+)
+
+// The number of times to do a roundtrip for each benchmark.
+// A high number favors gob more because it amortizes the cost
+// of encoding the type
+const count = 1000
```

Comment:
Some basic benchmarks comparing proto encoding to gob encoding. With the example search result, which was pulled from real data, proto manages to beat gob even as the count increases towards infinity. This isn't true for all payloads though, so it'll be worth looking into this more closely with production data.